r-package/grf/R/survival_forest.R

```diff
@@ -0,0 +1,206 @@
+  Y.relabeled <- findInterval(Y, failure.times, rightmost.closed = FALSE, all.inside = FALSE, left.open = FALSE)
+
+  data <- create_train_matrices(X, outcome = Y.relabeled, sample.weights = sample.weights, censor = D)
+  args <- list(num.trees = num.trees,
+               clusters = clusters,
+               samples.per.cluster = samples.per.cluster,
+               sample.fraction = sample.fraction,
+               mtry = mtry,
+               min.node.size = min.node.size,
+               honesty = honesty,
+               honesty.fraction = honesty.fraction,
+               honesty.prune.leaves = honesty.prune.leaves,
+               alpha = alpha,
+               num.failures = length(failure.times),
+               compute.oob.predictions = compute.oob.predictions,
+               num.threads = num.threads,
+               seed = seed)
+
+  forest <- do.call.rcpp(survival_train, c(data, args))
+  class(forest) <- c("survival_forest", "grf")
+  forest[["X.orig"]] <- X
+  forest[["Y.orig"]] <- Y
+  forest[["Y.relabeled"]] <- Y.relabeled
+  forest[["D.orig"]] <- D
+  forest[["sample.weights"]] <- sample.weights
+  forest[["clusters"]] <- clusters
+  forest[["equalize.cluster.weights"]] <- equalize.cluster.weights
+  forest[["has.missing.values"]] <- has.missing.values
+  forest[["failure.times"]] <- failure.times
+
+  forest
+}
+
+#' Predict with a survival forest forest
+#'
+#' Gets estimates of the conditional survival function S(t, x) using a trained surival forest (estimated using
+#' Kaplan-Meier).
+#'
+#' @param object The trained forest.
+#' @param newdata Points at which predictions should be made. If NULL, makes out-of-bag
+#'                predictions on the training set instead (i.e., provides predictions at
+#'                Xi using only trees that did not use the i-th training example). Note
+#'                that this matrix should have the number of columns as the training
+#'                matrix, and that the columns must appear in the same order.
+#' @param num.threads Number of threads used in training. If set to NULL, the software
+#'                    automatically selects an appropriate amount.
+#' @param ... Additional arguments (currently ignored).
+#'
+#' @return Vector of predictions.
+#'
+#' @examples
+#' \dontrun{
+#' # Train a standard survival forest.
+#' n <- 100
+#' p <- 5
+#' X <- matrix(rnorm(n * p), n, p)
+#' latent.time <- -log(runif(n)) * exp(0.1 * X[, 1])
+#' censor.time <- rexp(n)
+#' Y <- pmin(latent.time, censor.time)
+#' D <- as.integer(latent.time <= censor.time)
+#' s.forest <- survival_forest(X, Y, D)
+#'
+#' # Predict using the forest.
+#' X.test <- matrix(0, 101, p)
+#' X.test[, 1] <- seq(-2, 2, length.out = 101)
+#' s.pred <- predict(s.forest, X.test)
+#'
+#' # Predict on out-of-bag training samples.
+#' s.pred <- predict(s.forest)
+#' }
+#'
+#' @method predict survival_forest
+#' @export
+predict.survival_forest <- function(object, newdata = NULL, num.threads = NULL, ...) {
```

Comment:
This would only be for the case `newdata=X.new`? (like `predict(sf, X.test, failure.times = c(1, 5, 20))`) Yes it is possible, I just have to think a bit more.

Reply:
I'll leave this for now then, we can add it later if needed!

Reply:
Sounds good